### PR TITLE
recover WSV on startup

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
Signed-off-by: Alexey Chernyshov <alexey.n.chernyshov@gmail.com>

### Description of the Change
Recover WSV on startup of iroha to ensure it is consistent.

### Benefits
WSV is consistent

### Possible Drawbacks 
slowdown the startup

### Usage Examples or Tests *[optional]*
tests are included

### CODE REVIEW

File: irohad/ametsuchi/impl/storage_impl.hpp [hunk 1]
```diff
@@ -62,6 +62,12 @@ namespace iroha {
 
       virtual void dropStorage() override;
 
+      /**
+       * Recover WSV (World State View).
+       * Drop storage and apply blocks one by one.
+       */
+      virtual void recoverWSV() override;
```
Comment (l4l) [inline irohad/ametsuchi/impl/storage_impl.hpp]: Status on return will be useful, I guess bool is more than enough

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 2]
```diff
@@ -179,6 +179,38 @@ DROP TABLE IF EXISTS index_by_id_height_asset;
...
+                blocks.begin(), blocks.end(), [&mutableStorage](auto block) {
+                  mutableStorage.value->apply(
+                      block,
+                      [](const auto &block, auto &query, const auto &hash) {
+                        return true;
+                      });
+                });
+            commit(std::move(mutableStorage.value));
+          },
+          [&](iroha::expected::Error<std::string> &error) {
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply irohad/ametsuchi/impl/storage_impl.cpp]: It is for capture of class member `log_`.

File: irohad/main/irohad.cpp [hunk 3]
```diff
@@ -146,7 +146,14 @@ int main(int argc, char *argv[]) {
     irohad.storage->insertBlock(block.value());
     log->info("Genesis block inserted, number of transactions: {}",
               block.value().transactions.size());
+  } else {
+    // Recover VSW from the existing ledger to be sure it is consistent
+    if (not irohad.storage->recoverWSV()) {
+      log->error("Failed to recover WSV");
```
Comment (lebdron) [inline irohad/main/irohad.cpp]: At the moment, error is displayed at least two times - `match()` in `recoverWSV`, and here. Having a result and `match()` call here will be more consistent for logging.

File: test/module/irohad/ametsuchi/ametsuchi_fixture.hpp [hunk 4]
```diff
@@ -63,6 +62,14 @@ namespace iroha {
         } catch (const pqxx::broken_connection &e) {
           FAIL() << "Connection to PostgreSQL broken: " << e.what();
         }
+
+        auto storageResult = StorageImpl::create(block_store_path, pgopt_);
```
Comment (lebdron) [inline test/module/irohad/ametsuchi/ametsuchi_fixture.hpp]: 👍

File: test/module/irohad/ametsuchi/ametsuchi_test.cpp [hunk 5]
```diff
@@ -866,3 +803,53 @@ TEST_F(AmetsuchiTest, FindTxByHashTest) {
...
+      iroha::model::generators::BlockGenerator().generateGenesisBlock(
+          0, {genesis_tx});
+  apply(storage, genesis_block);
+
+  auto res = storage->getWsvQuery()->getDomain("test");
+  EXPECT_TRUE(res);
+
+  // spoil WSV
+  pqxx::work txn(*connection);
+  txn.exec(R"(
```
Comment (lebdron) [inline test/module/irohad/ametsuchi/ametsuchi_test.cpp]: Maybe it is enough to drop `domain` for example?

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 6]
```diff
@@ -179,6 +179,41 @@ DROP TABLE IF EXISTS index_by_id_height_asset;
...
+
+      // get all blocks starting from the genesis
+      std::vector<model::Block> blocks;
+      getBlockQuery()->getBlocksFrom(1).as_blocking().subscribe(
+          [&blocks](auto block) { blocks.push_back(block); });
+
+      dropStorage();
+
+      // apply blocks one by one
+      auto storageResult = createMutableStorage();
```
[the remark this answers is not in the document]
Comment (Alexey-N-Chernyshov) [reply irohad/ametsuchi/impl/storage_impl.cpp]: Is it possible to mock template functions with gmock?

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 7]
```diff
@@ -179,6 +204,20 @@ DROP TABLE IF EXISTS index_by_id_height_asset;
       block_store_->dropAll();
     }
 
+    bool StorageImpl::recoverWsv() {
+      log_->info("[Recover WSV] => start");
+
+      // get all blocks starting from the genesis
+      std::vector<model::Block> blocks;
+      getBlockQuery()->getBlocksFrom(1).as_blocking().subscribe(
+          [&blocks](auto block) { blocks.push_back(std::move(block)); });
```
Comment (l4l) [inline irohad/ametsuchi/impl/storage_impl.cpp]: emplace_back

Comment (lebdron) [reply irohad/ametsuchi/impl/storage_impl.cpp]:
```C++
void push_back( T&& value );
```
No need to call constructor explicitly.

File: irohad/ametsuchi/impl/wsv_restorer_impl.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "wsv_restorer_impl.hpp"
+
+#include <vector>
+
```
Comment (lebdron) [inline irohad/ametsuchi/impl/wsv_restorer_impl.cpp]: It is not required to explicitly call copy constructor with `emplace_back`. It is also possible to use move here: `blocks.push_back(std::move(*old_block));`

File: irohad/main/application.cpp [hunk 9]
```diff
@@ -94,17 +101,13 @@ void Irohad::initStorage() {
       [&](expected::Value<std::shared_ptr<ametsuchi::StorageImpl>> &_storage) {
         storage = _storage.value;
       },
-      [&](expected::Error<std::string> &error) {
-        log_->error(error.error);
```
Comment (lebdron) [inline irohad/main/application.cpp]: -> WSV

File: irohad/main/irohad.cpp [hunk 10]
```diff
@@ -20,6 +20,7 @@
 #include <csignal>
 #include <fstream>
 #include <thread>
+#include "common/result.hpp"
 #include "crypto/keys_manager_impl.hpp"
 #include "main/application.hpp"
 #include "main/iroha_conf_loader.hpp"
```
Comment (lebdron) [inline irohad/main/irohad.cpp]: Not used in this file.

**Comment by lebdron** [comment]
|`test/integration/pipeline/transfer_asset_inter_domain_test.cpp:84-85`|
|---|
`init` already contains `restoreWsv` call, so this should not be required.

File: irohad/ametsuchi/impl/storage_impl.hpp [hunk 11]
```diff
@@ -57,8 +57,21 @@ namespace iroha {
...
+       * Raw insertion of block without validation
+       * @param blocks - block for insertion
+       * @return true if inserted
+       */
       virtual bool insertBlock(model::Block block) override;
 
+      /**
+       * Raw insertion of blocks without validation
+       * @param blocks - collection of blocks for insertion
+       * @return true if inserted
```
Comment (l4l) [inline irohad/ametsuchi/impl/storage_impl.hpp]: if all blocks are inserted

File: irohad/ametsuchi/impl/storage_impl.hpp [hunk 12]
```diff
@@ -57,8 +57,21 @@ namespace iroha {
       expected::Result<std::unique_ptr<MutableStorage>, std::string>
       createMutableStorage() override;
 
+      /**
+       * Raw insertion of block without validation
```
Comment (l4l) [inline irohad/ametsuchi/impl/storage_impl.hpp]: Does that "Raw" introduce some (useful) info?

File: irohad/ametsuchi/wsv_restorer.hpp [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "common/result.hpp"
+
+namespace iroha {
+  namespace ametsuchi {
+
+    class Storage;
+
+    /**
+     * Interface for class that restores World State View from storage.
```
Comment (l4l) [inline irohad/ametsuchi/wsv_restorer.hpp]: Interface for World State View restoring from the storage

File: irohad/ametsuchi/wsv_restorer.hpp [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    class WsvRestorer {
+     public:
+      virtual ~WsvRestorer() = default;
+
+      /**
+       * Recover WSV (World State View).
+       * @return void on success, otherwise error string
```
Comment (l4l) [inline irohad/ametsuchi/wsv_restorer.hpp]: `@param storage` missing?

File: test/module/irohad/ametsuchi/ametsuchi_test.cpp [hunk 15]
```diff
@@ -24,16 +24,21 @@
...
 #include "ametsuchi/impl/postgres_wsv_query.hpp"
-#include "ametsuchi/impl/storage_impl.hpp"
+#include "ametsuchi/impl/wsv_restorer_impl.hpp"
 #include "ametsuchi/mutable_storage.hpp"
+#include "builders/protobuf/block.hpp"
+#include "builders/protobuf/transaction.hpp"
 #include "common/byteutils.hpp"
+#include "cryptography/hash_providers/sha3_256.hpp"
 #include "framework/test_subscriber.hpp"
 #include "model/account.hpp"
```
Comment (l4l) [inline test/module/irohad/ametsuchi/ametsuchi_test.cpp]: Can't find usage of `model::Account` in the file, so the include seems unused. Remove the rest
Remove the rest of unused headers plz (may help `using namespace iroha::model` removal)

Comment (Alexey-N-Chernyshov) [reply test/module/irohad/ametsuchi/ametsuchi_test.cpp]: needed for wsv->getAccount()

Comment (Alexey-N-Chernyshov) [reply test/module/irohad/ametsuchi/ametsuchi_test.cpp]: But I removed a bunch of other includes.

Comment (l4l) [reply test/module/irohad/ametsuchi/ametsuchi_test.cpp]: Exact type isn't used, it basically inherited from getAccount type

**Pull request merged**